File: frontend/scripts/apply-education-ja-translations.mjs
#!/usr/bin/env node
/** Fill Japanese translations for education wisdom articles (p1–p6). */

import { readFileSync, writeFileSync, existsSync } from "fs";
import { dirname, join } from "path";
import { fileURLToPath } from "url";

const __dirname = dirname(fileURLToPath(import.meta.url));
const EDU = join(__dirname, "../src/lib/education");
const CACHE = join(__dirname, "ja-education-translations.json");

const FILES = [
  "p1-content.ts",
  "p2-content.ts",
  "p3-content.ts",
  "p4-content.ts",
  "p5-content.ts",
  "p6-content.ts",
];

const PAIR_RE =
  /en:\s*(?<en>`(?:\\.|[^`\\])*`|"(?:\\.|[^"\\])*")\s*,?\s*ja:\s*(?<ja>`(?:\\.|[^`\\])*`|"(?:\\.|[^"\\])*")/gs;

function unquote(raw) {
  const q = raw[0];
  const body = raw.slice(1, -1);
  if (q === "`") {
    return body
      .replace(/\\`/g, "`")
      .replace(/\\\$\{/g, "${")
      .replace(/\\\\/g, "\\");
  }
  return body.replace(/\\"/g, '"').replace(/\\\\/g, "\\");
}

function quote(text, style) {
  if (style === "`") {
    const escaped = text
      .replace(/\\/g, "\\\\")
      .replace(/`/g, "\\`")
      .replace(/\$\{/g, "\\${");
    return `\`${escaped}\``;
  }
  const escaped = text.replace(/\\/g, "\\\\").replace(/"/g, '\\"');
  return `"${escaped}"`;
}

async function translateText(text, attempt = 0) {
  if (!text.trim()) return text;
  const url = new URL("https://translate.googleapis.com/translate_a/single");
  url.searchParams.set("client", "gtx");
  url.searchParams.set("sl", "en");
  url.searchParams.set("tl", "ja");
  url.searchParams.set("dt", "t");
  url.searchParams.set("q", text);

  const res = await fetch(url);
  if (!res.ok) {
    if (attempt < 3) {
      await sleep(500 * (attempt + 1));
      return translateText(text, attempt + 1);
    }
    throw new Error(`HTTP ${res.status}`);
  }
  const data = await res.json();
  return data[0].map((part) => part[0]).join("");
}

function loadCache() {
  if (!existsSync(CACHE)) return {};
  return JSON.parse(readFileSync(CACHE, "utf8"));
}

function saveCache(cache) {
  writeFileSync(CACHE, JSON.stringify(cache, null, 2) + "\n", "utf8");
}

function collectPairs(content) {
  const pairs = [];
  for (const match of content.matchAll(PAIR_RE)) {
    const enRaw = match.groups.en;
    const jaRaw = match.groups.ja;
    const en = unquote(enRaw);
    const ja = unquote(jaRaw);
    if (en === ja) pairs.push({ enRaw, jaRaw, en });
  }
  return pairs;
}

function applyFile(path, cache) {
  let content = readFileSync(path, "utf8");
  let updated = 0;
  for (const { enRaw, jaRaw, en } of collectPairs(content)) {
    const jaText = cache[en];
    if (!jaText || jaText === en) continue;
    const style = jaRaw[0];
    const newJaRaw = quote(jaText, style);
    if (newJaRaw === jaRaw) continue;
    const patterns = [
      `en: ${enRaw}, ja: ${jaRaw}`,
      `en: ${enRaw},\n      ja: ${jaRaw}`,
      `en: ${enRaw},\n          ja: ${jaRaw}`,
      `en: ${enRaw},\n        ja: ${jaRaw}`,
    ];
    let replaced = false;
    for (const needle of patterns) {
      if (content.includes(needle)) {
        content = content.replace(needle, needle.replace(`ja: ${jaRaw}`, `ja: ${newJaRaw}`));
        replaced = true;
        break;
      }
    }
    if (!replaced) continue;
    updated += 1;
  }
  if (updated) writeFileSync(path, content, "utf8");
  return updated;
}

function sleep(ms) {
  return new Promise((resolve) => setTimeout(resolve, ms));
}

async function main() {
  const cache = loadCache();
  const unique = new Set();

  for (const name of FILES) {
    const content = readFileSync(join(EDU, name), "utf8");
    for (const { en } of collectPairs(content)) {
      if (!cache[en] || cache[en] === en) unique.add(en);
    }
  }

  console.log(`Need to translate ${unique.size} unique strings`);
  let done = 0;
  for (const en of [...unique].sort((a, b) => a.length - b.length)) {
    try {
      cache[en] = await translateText(en);
    } catch (err) {
      console.error(`Failed: ${en.slice(0, 80)}... (${err.message})`);
      continue;
    }
    done += 1;
    if (done % 25 === 0) {
      saveCache(cache);
      console.log(`Translated ${done}/${unique.size}`);
    }
    await sleep(120);
  }

  saveCache(cache);

  let total = 0;
  for (const name of FILES) {
    const count = applyFile(join(EDU, name), cache);
    console.log(`${name}: updated ${count} ja fields`);
    total += count;
  }
  console.log(`Done. Applied ${total} replacements.`);
}

main().catch((err) => {
  console.error(err);
  process.exit(1);
});

File: frontend/scripts/apply-education-ja-translations.py
#!/usr/bin/env python3
"""Fill Japanese translations for education wisdom articles (p1–p6)."""

from __future__ import annotations

import json
import re
import sys
import time
from pathlib import Path

try:
    from deep_translator import GoogleTranslator
except ImportError:
    print("Install deep-translator: pip install deep-translator", file=sys.stderr)
    raise

DIR = Path(__file__).parent
EDU = DIR.parent / "src/lib/education"
CACHE = DIR / "ja-education-translations.json"

FILES = [
    "p1-content.ts",
    "p2-content.ts",
    "p3-content.ts",
    "p4-content.ts",
    "p5-content.ts",
    "p6-content.ts",
]

PAIR_RE = re.compile(
    r"en:\s*(?P<en>`(?:\\.|[^`\\])*`|\"(?:\\.|[^\"\\])*\")\s*,\s*ja:\s*(?P<ja>`(?:\\.|[^`\\])*`|\"(?:\\.|[^\"\\])*\")",
    re.DOTALL,
)


def unquote(raw: str) -> str:
    q = raw[0]
    body = raw[1:-1]
    if q == "`":
        return (
            body.replace("\\`", "`")
            .replace("\\${", "${")
            .replace("\\\\", "\\")
        )
    return body.replace('\\"', '"').replace("\\\\", "\\")


def quote(text: str, style: str) -> str:
    if style == "`":
        escaped = text.replace("\\", "\\\\").replace("`", "\\`").replace("${", "\\${")
        return f"`{escaped}`"
    escaped = text.replace("\\", "\\\\").replace('"', '\\"')
    return f'"{escaped}"'


def translate_text(text: str, translator: GoogleTranslator) -> str:
    if not text.strip():
        return text
    # Preserve markdown markers and footnotes by translating in chunks if needed.
    chunk_size = 4500
    if len(text) <= chunk_size:
        return translator.translate(text)

    parts: list[str] = []
    start = 0
    while start < len(text):
        end = min(start + chunk_size, len(text))
        if end < len(text):
            split = text.rfind(". ", start, end)
            if split > start + 500:
                end = split + 1
        parts.append(translator.translate(text[start:end]))
        start = end
        time.sleep(0.15)
    return "".join(parts)


def load_cache() -> dict[str, str]:
    if CACHE.exists():
        return json.loads(CACHE.read_text(encoding="utf-8"))
    return {}


def save_cache(cache: dict[str, str]) -> None:
    CACHE.write_text(json.dumps(cache, ensure_ascii=False, indent=2) + "\n", encoding="utf-8")


def collect_pairs(content: str) -> list[tuple[str, str, str]]:
    pairs: list[tuple[str, str, str]] = []
    for match in PAIR_RE.finditer(content):
        en_raw, ja_raw = match.group("en"), match.group("ja")
        en = unquote(en_raw)
        ja = unquote(ja_raw)
        if en == ja:
            pairs.append((en_raw, ja_raw, en))
    return pairs


def apply_file(path: Path, cache: dict[str, str]) -> int:
    content = path.read_text(encoding="utf-8")
    updated = 0
    for en_raw, ja_raw, en in collect_pairs(content):
        ja_text = cache.get(en)
        if not ja_text or ja_text == en:
            continue
        style = ja_raw[0]
        new_ja_raw = quote(ja_text, style)
        if new_ja_raw == ja_raw:
            continue
        content = content.replace(f"en: {en_raw}, ja: {ja_raw}", f"en: {en_raw}, ja: {new_ja_raw}", 1)
        updated += 1
    if updated:
        path.write_text(content, encoding="utf-8")
    return updated


def main() -> None:
    cache = load_cache()
    translator = GoogleTranslator(source="en", target="ja")

    unique: set[str] = set()
    for name in FILES:
        path = EDU / name
        for _, _, en in collect_pairs(path.read_text(encoding="utf-8")):
            if en not in cache or cache[en] == en:
                unique.add(en)

    print(f"Need to translate {len(unique)} unique strings")
    done = 0
    for en in sorted(unique, key=len):
        try:
            cache[en] = translate_text(en, translator)
        except Exception as exc:  # noqa: BLE001
            print(f"Failed ({exc}): {en[:80]}...")
            continue
        done += 1
        if done % 25 == 0:
            save_cache(cache)
            print(f"Translated {done}/{len(unique)}")
        time.sleep(0.12)

    save_cache(cache)

    total = 0
    for name in FILES:
        count = apply_file(EDU / name, cache)
        print(f"{name}: updated {count} ja fields")
        total += count
    print(f"Done. Applied {total} replacements. Cache: {CACHE}")


if __name__ == "__main__":
    main()
